examples-draft/SimpleExampleWF_Max.tla

```diff
@@ -0,0 +1,197 @@
+--------------------------- MODULE SimpleExampleWF_Max -----------------------
+(*
+That's an example of a proof of Spec => <>P. Such proof has to be based on
+a fairness condition. The general idea is that we have to show that either
+<>P is true, or an action is enabled that will make it true later.
+
+The example here defines an operator Max, then an algorithm that
+"calculates" the maximum in several steps (MaxAsFSM) and finally proves
+that the algorithm eventually will give the same result as the
+operator Max (a kind of refinement). In this way we model a non-atomic
+implementation of an operation defined in a declarative way.
+
+In the example SimpleExampleWF_Flag the variable was a boolean, thus we
+managed to prove the P \/ ENABLED A just from the definitions of P and A.
+Here we cannot do that and induction is needed to show P \/ ENABLED A as
+an inductive invariant.
+*)
+EXTENDS Naturals, TLAPS
+
+Max(a, b) ==
+  IF a > b THEN a ELSE b 
+
+-----------------------------------------------------------------------------
+
+---------- MODULE MaxAsFSM ---------------
+EXTENDS Naturals
+CONSTANTS a, b
+ASSUME Assms == a \in Nat /\ b \in Nat
+VARIABLES x
+Null == CHOOSE n : n \notin Nat
+Init == x = Null
+Next ==
+ /\ x = Null
+ /\ \/ a >  b /\ x' = a
+    \/ a =< b /\ x' = b
+Live == WF_x(Next)
+Spec == Init /\ [][Next]_x /\ Live
+TypeOK == x \in Nat \cup {Null}
+==========================================
+CONSTANTS mA, mB
+VARIABLE mX
+ASSUME mAssms == mA \in Nat /\ mB \in Nat
+m == INSTANCE MaxAsFSM WITH
+  a <- mA,
+  b <- mB,
+  x <- mX
+
+mIsMax == mX = Max(mA, mB)
+
+LEMMA mEnabledNext == mX = m!Null => ENABLED <<m!Next>>_mX
```

Comment:
@muenchnerkindl 
Was playing with this proof to understand why it works. Found a case making the TLAPS to hang:
```
LEMMA mEnabledNext == mX = m!Null => ENABLED <<m!Next>>_mX
  (* First we show that stuttering steps are not allowed by the action. *)
  <1>1. m!Next => <<m!Next>>_mX
        BY m!Assms DEF m!Next, m!Null
  (* This allows us to drop ... /\ mX # mX' in the goal. *)
  <1>2. SUFFICES ASSUME mX = m!Null PROVE ENABLED (m!Next /\ mX' = mA)
        BY <1>1, m!Assms, ExpandENABLED DEF m!Next, m!Null
  (* Finally by replacing all the equations involving primed variables
     with TRUE (by ENABLEDrewrites) we prove our goal. *)
  <1>4. QED BY <1>2, ENABLEDrewrites, m!Assms DEF m!Next
```
Notice the `/\ mX' = mA` in <1>2. The expression itself is not meaningful in this context, but it breaks the prover somehow.
The TLAPS doesn't crash, but hangs (was waiting for several minutes) and consumes CPU with no visible progress.